### PR TITLE
Refactor controllers to decouple service instantiation

### LINKED ISSUE #1047: Refactor controller to decouple service instantiaition
/kind cleanup

**Describe the solution you'd like**
Right now, in controllers, the reconciliation service is instantiated directly inside the `reconcileNormal` and `reconcileDelete` methods like [this](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/5ce2cc61a6862852cd7ed7229757422c6f6b7dfb/controllers/azuremachine_controller.go#L266). This tight coupling makes it difficult to write tests since we cannot mock the behavior of this service as experienced in #1042 . So, we want to remove this coupling and inject this as a dependency in the reconciler instance. The following are a list of changes that should be done as part of this refactor:
1. Add a factory in reconciler to create service instances
2. Change instantiation logic to use the factory
3. Extract an interface out of the service instance (this will be similar to `azure.Service` interface)

**Anything else you would like to add:**
The changes will look something like this:

```
type AzureMachineReconciler struct {
	client.Client
	Log              logr.Logger
	Recorder         record.EventRecorder
	ReconcileTimeout time.Duration
	azureMachineServiceFactory func(*scope.MachineScope, *scope.ClusterScope) azureMachineService
}

// inject it using the factory 
func (r *AzureMachineReconciler) ctx context.Context, machineScope *scope.MachineScope, clusterScope *scope.ClusterScope) (reconcile.Result, error) {
 
ams := r.azureMachineServiceFactory(machineScope, clusterScope)

}

type Service interface {
   func Reconcile(ctx) error
   func Delete(ctx) error

}

``` 

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by shysank**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

This is essentially rinse and repeat of #1053 for all controllers

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1047 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Refactor controllers to decouple service instantiation
```

### CODE REVIEW

**Comment by shysank** [comment]
@CecileRobertMichon @devigned Do you think it'd make it easier to review if I just merged this with #1053 ?

**Comment by CecileRobertMichon** [comment]
I just approved #1053, let's just merge them separately

File: controllers/azurecluster_reconciler.go
Comment (CecileRobertMichon) [inline controllers/azurecluster_reconciler.go:40]: thanks for fixing that, I've been meaning to do it for a while

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/assign @devigned

File: controllers/azurecluster_reconciler.go
Comment (devigned) [reply controllers/azurecluster_reconciler.go:40]: I think this may overlap with https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1075

Comment (shysank) [reply controllers/azurecluster_reconciler.go:40]: Yeah, one of us has to rebase. Happy to do it after #1075 gets merged.

/cc @surajrana93

**Comment by devigned** [comment]
@shysank, my apologies. I just realized I've been holding this up. Please rebase and we'll get this merged.

**Comment by shysank** [comment]
@devigned I've rebased with master, PTAL.

**Review by devigned — COMMENTED** [review]
The changes look good. I'm hesitant to approve until #1075 since this will create a lot of naming churn due to Serivce / Reconciler interface refactor.

We may need to be patient here as many folks are away for holiday.

**Comment by CecileRobertMichon** [comment]
@shysank if you don't mind rebasing this we'll get it pulled in. #1075 has a lot of conflicts already...

**Review by CecileRobertMichon — APPROVED** [review]
/approve

**Pull request merged**
